org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/utils/MarkdownConverter.java

```diff
@@ -0,0 +1,76 @@
+/*******************************************************************************
+ * Copyright (c) 2016-2017 Red Hat Inc. and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Red Hat, Inc. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.lsp4xml.utils;
+
+import java.lang.reflect.Field;
+import java.util.logging.Logger;
+import java.util.regex.Pattern;
+
+import com.overzealous.remark.Options;
+import com.overzealous.remark.Options.Tables;
+import com.overzealous.remark.Remark;
+
+import org.apache.commons.lang3.StringEscapeUtils;
+import org.jsoup.safety.Cleaner;
+import org.jsoup.safety.Whitelist;
+
+/**
+ * Converts HTML content into Markdown equivalent.
+ *
+ * @author Fred Bricon
+ */
+public class MarkdownConverter {
```

Comment:
Here a version of markdown converter (written in few minutes, so I think we can improve it again to support other HTML elements) which uses the XMLScanner. All tests are passed with this converter.

```java
/*******************************************************************************
 * Copyright (c) 2016-2017 Red Hat Inc. and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     Red Hat, Inc. - initial API and implementation
 *******************************************************************************/
package org.eclipse.lsp4xml.utils;

import java.util.logging.Logger;
import java.util.regex.Pattern;

import org.apache.commons.lang3.StringEscapeUtils;
import org.eclipse.lsp4xml.dom.parser.Scanner;
import org.eclipse.lsp4xml.dom.parser.TokenType;
import org.eclipse.lsp4xml.dom.parser.XMLScanner;

/**
 * Converts HTML content into Markdown equivalent.
 *
 * @author Fred Bricon
 */
public class MarkdownConverter {

	private static final Logger LOGGER = Logger.getLogger(MarkdownConverter.class.getName());

	// Pattern looking for any form of tag eg: <head>
	private static final Pattern htmlPattern = Pattern.compile("[^`]*<[a-z][\\s\\S]*>[^`]*");

	private MarkdownConverter() {
		// no public instanciation
	}

	public static String convert(String html) {
		if (!htmlPattern.matcher(html).matches()) {
			return StringEscapeUtils.unescapeXml(html); // is not html so it can be returned
		}
		return doConvert(html);
	}

	private static String doConvert(String html) {
		StringBuilder markdown = new StringBuilder();
		Scanner scanner = XMLScanner.createScanner(html);
		TokenType token = scanner.scan();
		while (token != TokenType.EOS) {
			switch (token) {
			case Content:
				markdown.append(scanner.getTokenText().replaceAll("&lt;", "<").replaceAll("&gt;", ">"));
				break;
			case StartTag: {
				String tag = scanner.getTokenText();
				if ("code".equals(tag)) {
					markdown.append("`");
				} else if ("br".equals(tag)) {
					markdown.append("\n");
				} else if ("b".equals(tag) || "strong".equals(tag)) {
					markdown.append("**");
				}
				break;
			}
			case EndTag: {
				String tag = scanner.getTokenText();
				if ("code".equals(tag)) {
					markdown.append("`");
				} else if ("b".equals(tag) || "strong".equals(tag)) {
					markdown.append("**");
				}
				break;
			}
			}
			token = scanner.scan();
		}

		return markdown.toString();
	}

}

```